---
title: "Intro to Epigenomics"
author: Wade Boohar
date: 11/03/24
updated: 03/07/24
---


```{r setup}
 knitr::opts_knit$set(root.dir = normalizePath("/home1/leonjche/490_cluster/analysis_data"))
```

Package Download and Data-cleaning
```{r}
if (!require("sesameData", quietly = TRUE))
BiocManager::install("sesameData")

if (!require("sesame", quietly = TRUE))
BiocManager::install("sesame")

if (!require("limma", quietly = TRUE))
BiocManager::install("limma")
```


Load in all necessary packages
```{r}
library(TCGAbiolinks)
library(sesame)
library(sesameData)
library(limma)
library(ggplot2)
library(DESeq2)
```

```{r}
methylation_clinical <- read.csv("/project/rohs_1070/analysis_data/brca_methylation_clinical.csv", row.names = 1)

cpg_sites <- read.csv("/project/rohs_1070/analysis_data/brca_cpg_sites.csv", row.names = 1)


#install.packages("data.table")
library(data.table)

betas <- fread('/project/rohs_1070/analysis_data/brca_methylation_betas.csv', showProgress = TRUE)

betas <- as.data.frame(betas)
rownames(betas) <- betas[[1]]
betas[[1]] <- NULL
```

Change to metastatic status instead of age
(1) Naive Differential Methylation
```{r}
#masking out NAs
na_mask <- !is.na(methylation_clinical$age_at_diagnosis)
methylation_clinical <- methylation_clinical[na_mask,]
betas_clean <- betas[,na_mask]

median <- median(methylation_clinical$age_at_diagnosis)

methylation_clinical$age_category <- ifelse(methylation_clinical$age_at_diagnosis >= median, "old", "young")

#fitting linear models using a "target value"
young_mask <- methylation_clinical$age_category == "young"

methylation_clinical$ages <- young_mask

mval <- t(apply(betas_clean, 1, function(x) log2(x/(1-x)))) 
#mvalue is another statistic for methylation, centered at 0 and ranges from -1 to 1

design <- model.matrix(~ages, data = methylation_clinical)
fit <- lmFit(mval, design)
fit2 <- eBayes(fit)
```

```{r}
#Extracting model into dataframe
dat <- data.frame(foldchange = fit[["coefficients"]][,2], logPvalue =  -log10(p.adjust(fit2[["p.value"]][,2],method='BY')), geneName = cpg_sites$gene)
dat$threshold <- as.factor(abs(dat$foldchange) < 1)

#Visualization
cols <- c("TRUE" = "grey", "FALSE" = "blue")
volcano_plot1 <- ggplot(data=dat, aes(x=foldchange, y = logPvalue, color=threshold)) +
  geom_point(alpha=.2, size=0.6) +
  scale_colour_manual(values = cols) +
  geom_vline(xintercept = 1, colour="#990000", linetype="dashed") + 
  geom_vline(xintercept = - 1, colour="#990000", linetype="dashed") +
  geom_hline(yintercept = 2, colour = "#990000", linetype="dashed") +
  theme(legend.position="none") +
  xlab("Fold Change") +
  ylab("-log10 p value") +
  theme_bw() +
  theme(legend.position = "none")


ggsave(
  filename = "/home1/leonjche/490_cluster/week08_Epigenomics/volcano_plot_1.png",
  plot = volcano_plot1,
  width = 8,   # inches
  height = 6,
  dpi = 300
)
```


(2) Direct comparison of methylation status to transcriptional activity

#...INSERT DESeq2 Stuff here to generate 'results'...
```{r}
library(EnhancedVolcano)
rna_clinical <- read.csv("brca_rna_clinical_data.csv")
unique(rna_clinical$definition)

rna_genes <- read.csv("brca_rna_gene_data.csv")
rna_counts <- read.csv("brca_rna_count_data.csv")
#Clean up and make sure types are correct
rna_clinical$definition <- droplevels(factor(rna_clinical$definition))

#Covariate
rna_clinical$age_at_diagnosis <- as.numeric(rna_clinical$age_at_diagnosis)

#Remove samples with NA in either variable
rna_clinical <- rna_clinical[
  !is.na(rna_clinical$definition) & !is.na(rna_clinical$age_at_diagnosis),
]

#give proper colnames and rownames
rownames(rna_clinical) <- rna_clinical$X
rna_clinical <- rna_clinical[,-1]

rownames(rna_counts) <- rna_counts$X
rna_counts <- rna_counts[,-1]
#naming convention
colnames(rna_counts)       <- gsub("\\.", "-", colnames(rna_counts))
rownames(rna_clinical) <- gsub("\\.", "-", rownames(rna_clinical))


#prefiltering genes

keep_samples <- rowSums(rna_counts) >= 1000
filtered_rna_counts <- rna_counts[keep_samples,]
dim(filtered_rna_counts)

common <- intersect(colnames(filtered_rna_counts), rownames(rna_clinical))
filtered_rna_counts <- filtered_rna_counts[, common, drop = FALSE]
rna_clinical <- rna_clinical[common, , drop = FALSE]
dim(filtered_rna_counts)
```

```{r}
#~covariate 1 + covariate 2 + variable of interest
dds <- DESeqDataSetFromMatrix(countData = filtered_rna_counts,
                              colData = rna_clinical,
                              design= ~age_at_diagnosis + definition)
```

```{r}
dds_obj <- DESeq(dds)

resultsNames(dds_obj)

results <- results(dds_obj, name = "definition_Solid.Tissue.Normal_vs_Metastatic")

results <- data.frame(results)

results_sig <- subset(results,
                      !is.na(padj) &
                      padj < 0.05 &
                      abs(log2FoldChange) >= 1)
```

```{r}
results$padj[is.na(results$padj)] <- 1
results$'-log10(padj)' <- -log10(results$padj)
results$gene_id <- rownames(results)
results$gene_name <- rna_genes$gene_name[match(results$gene_id, rna_genes$gene_id)]


#-log padj isnt numeric
for (cn in c("padj","pvalue","log2FoldChange","lfcSE","stat","baseMean")) {
  if (cn %in% names(results)) {
    results[[cn]] <- suppressWarnings(as.numeric(gsub(",", "", as.character(results[[cn]]))))
  }
}N
```

```{r}

install.packages(AnnotationDbi)


ens <- sub("\\..*$", "", rownames(results))
install.packages("org.Hs.eg.db")
  library(AnnotationDbi)
  library(org.Hs.eg.db)

map <- AnnotationDbi::select(org.Hs.eg.db,
                             keys    = unique(ens),
                             keytype = "ENSEMBL",
                             columns = "SYMBOL")
sym_lookup <- setNames(map$SYMBOL, map$ENSEMBL)

results$gene_name <- sym_lookup[ens]
missing <- is.na(results$gene_name)
results$gene_name[missing] <- ens[missing]

results$`-log10(padj)` <- -log10(results$padj)
results$`-log10(padj)`[!is.finite(results$`-log10(padj)`)]<- NA_real_
```

```{r}
volcano_plot <- EnhancedVolcano(results,
                  lab = results$gene_name,
                  x = 'log2FoldChange',
                  y = 'padj',
                  title = 'Sample Definition: Tumor vs Normal Tissue',
                  pointSize = 1.0,
                  labSize = 5.0)

library(ggplot2)
ggsave(
  filename = "/home1/leonjche/490_cluster/week08_Epigenomics/volcano_plot.png",
  plot = volcano_plot,
  width = 8,   # inches
  height = 6,
  dpi = 300
)
```


```{r}
#you can also try looking at "upregulated" or "hypermethylated" !
downregulated <- results[(results$log2FoldChange < 3), ]
downregulated <- downregulated$gene_name
hypomethylated <- dat[dat$foldchange < -1, 'geneName']
interest_genes <- intersect(downregulated, hypomethylated)
```


(Extra) Making Boxplots
```{r}
GENE<-"SCTR"

gene_counts_mask <- rna_genes$gene_name == GENE
gene_betas_mask <- cpg_sites$gene == GENE

rna_clinical_tumor <- rna_clinical$definition == "Primary solid Tumor"
methylation_clinical_tumor <- methylation_clinical$definition == "Primary solid Tumor"

rna_clinical_normal <- rna_clinical$definition == "Solid Tissue Normal"
methylation_clinical_normal <- methylation_clinical$definition == "Solid Tissue Normal"

rna_tumor <- as.numeric(rna_counts[gene_counts_mask, rna_clinical_tumor])
methylation_tumor <- (betas[gene_betas_mask, methylation_clinical_tumor])

rna_normal <- as.numeric(rna_counts[gene_counts_mask, rna_clinical_normal])
methylation_normal <- (betas[gene_betas_mask, methylation_clinical_normal])
```

```{r}
boxplot(rna_normal, rna_tumor, xlab='Group', ylab='Counts', names=c('Normal', 'Tumor'))
```
